### PR TITLE
Issue #1858: [UX] Project Browser should run some pre-flight checks

### PR DESCRIPTION
**Pull request by daggerhart**
Fixes issue: https://github.com/backdrop/backdrop-issues/issues/1858

This could be better, but I avoided changing any core installer files.  

Ideally, there would be a way to get the project-type-relative install path from the `Updater` instance without having to construct it.  To construct an `Updater` class, it requires a legitimate project name. 

Ex: `$updater = new ModuleUpdater( $legitmate_project_name );`

With a working instance of an Updater, I could call `$updater->getInstallDirectory()`, it would fail to find a real backdrop module path, and it would return with `/path/to/backdrop/modules`.

But since the constructor executes `Updater::getProjectTitle()`, and that method throws an exception when it fails, I couldn't find a great way to get a system-provided install path for a project-type.

Additionally, with a fully-functioning `Updater` instance, I could `Updater::prepareInstallDirectory()` for a more accurate test of whether or not the directory can be created during project installation.

Long story short, this PR leverages the coincidence of all project type install directories being `BACKDROP_ROOT."/".$type."s"`, and some simple mkdir/rmdir functions to test if the install directory is writable.  If it fails to make the directory, it provides a UI warning to the user.

As a simple internet test, it attempts to get the headers for example.com, and provides a UI error in the case of failure.

I don't think this is testable on the CI server. You would need to be able to change some file permissions, and somehow disconnect the site from example.com.

### CODE REVIEW

File: core/modules/installer/installer.browser.inc [hunk 1]
```diff
@@ -573,3 +573,36 @@ function installer_browser_check_zip_loaded() {
...
+    backdrop_set_message(t('The following directory is not writable by the server: "@path" You may have trouble installing @plural', array('@plural' => $plural, '@path' => $path,)), 'warning', false);
+  }
+}
+
+/**
+ *
+ * Checks if the backdrop contrib repo is accessible
+ */
+function installer_browser_check_repo_accessible() {
+  $response = backdrop_http_request('http://example.com/', array('method' => 'HEAD'));
```
Comment (jlfranklin) [inline core/modules/installer/installer.browser.inc:603]: Instead of hitting a third-party site with traffic, I think it would be better to test connectivity to the actual projects server. I think I see why you did this: there isn't any obvious path to test on the project server itself.  The root returns a 404, as does `robots.txt`.  I've opened backdrop-ops/backdropcms.org#436 to create such a file for this patch to test against.

File: core/modules/installer/installer.browser.inc [hunk 2]
```diff
@@ -573,3 +573,36 @@ function installer_browser_check_zip_loaded() {
...
+ *
+ * @param $type
+ *   Type of project to view. Must be one of 'module', 'theme' or 'layout'.
+ */
+function installer_browser_check_folder_writable($type) {
+  // luckily these are simple words to plural
+  $plural = "{$type}s";
+  $path = BACKDROP_ROOT."/{$plural}";
+  $test = $path.'/'.uniqid('test-');
+
```
Comment (jlfranklin) [inline core/modules/installer/installer.browser.inc:589]: The coding standards don't have [hard rules on quotes and string concatenation](https://api.backdropcms.org/php-standards#quotes), but they do say add spaces between the dot and and the strings being concatenated, and encourage "consistency within each module", and certainly within a single patch.   I think we should use the single-quote-and-dot approach for all three lines here.

File: core/modules/installer/installer.browser.inc [hunk 3]
```diff
@@ -573,3 +573,36 @@ function installer_browser_check_zip_loaded() {
...
+  if (backdrop_mkdir($test)){
+    backdrop_rmdir($test);
+  }
+  else {
```
Comment (jlfranklin) [inline core/modules/installer/installer.browser.inc:593]: Using `is_writable($path)` here would be more readable and avoids unnecessary writes to the filesystem.

File: core/modules/installer/installer.browser.inc [hunk 4]
```diff
@@ -573,3 +573,36 @@ function installer_browser_check_zip_loaded() {
...
   return $zip_loaded;
 }
+
+/**
+ *
+ * Checks if relevant system folder is writable
+ *
+ * @param $type
+ *   Type of project to view. Must be one of 'module', 'theme' or 'layout'.
+ */
```
Comment (jlfranklin) [inline core/modules/installer/installer.browser.inc:583]: Thanks for including documentation!  Removing the blank line at 538 makes the docblock more consistent with the rest of Backdrop.

File: core/modules/installer/installer.browser.inc [hunk 5]
```diff
@@ -573,3 +573,36 @@ function installer_browser_check_zip_loaded() {
...
+  }
+  else {
+    backdrop_set_message(t('The following directory is not writable by the server: "@path" You may have trouble installing @plural', array('@plural' => $plural, '@path' => $path,)), 'warning', false);
+  }
+}
+
+/**
+ *
+ * Checks if the backdrop contrib repo is accessible
+ */
```
Comment (jlfranklin) [inline core/modules/installer/installer.browser.inc:601]: Again, extra docblock whitespace (line 599).

**Review by jlfranklin — CHANGES_REQUESTED** [review]
Thanks for the patch!  There are a few items, mostly code style issues, that I think you need to fix.  Otherwise, it's a great patch.
